File: twoPointer/10025.cpp
#include <iostream>
using namespace std;
#define ll long long
int board[1000002];

int main(void){
    int n, k, g, x, end = 0;
    
    cin >> n >> k;
    for (int i = 0; i < n ;i++)
    {
        cin >> g >> x;
        board[x] = g;
        end = max(end, x);
    }
    if (k >= end)
    {
        int ans = 0;
        for (int i = 0; i <= end; i++)
            ans += board[i];
        cout << ans;
        return 0;
    }
    
    ll window = 2*k + 1;
    ll ice = 0;
    for (int i = 0; i < window; i++)
        ice += board[i];
        
    ll ans = ice;
    for (int i = window; i <= end; i++)
    {
        ice += board[i];
        ice -= board[i - window];
        ans = max(ice, ans);
    }
    cout << ans;
}

File: python/10025.py
n, k = map(int, input().split())
board = [0] * (1000002)

end = 0
for _ in range(n):
    g, x = map(int, input().split())
    board[x] = g       # board[x] 자리 얼음 g 그램
    end = max(end, x)  # 마지막 얼음이 들어있는 자리

window = 2*k + 1  # 움직일 수 있는 크기
ice = sum(board[:window])
ans = ice

for i in range(window, 1000001):  # 슬라이딩 윈도우 
    ice += board[i]
    ice -= board[i-window]
    ans = max(ans, ice)
print(ans)
